### DISCUSSION TITLE
Engine HAT digital input question

### DISCUSSION THREAD

**Discussion by sergei** [hatlabs/discussions#40] [Q&A] [answered]
I'm trying to send the pulses from water speed and wind speed sensors through the digital input pins.
The input looks like this:
![image](https://user-images.githubusercontent.com/138027/209414383-3573377a-8acc-4ff3-8d55-bfa024f79937.png)
If I connect it directly to GPIO 14 or 15 the  ESP32 counter function works just. However if hook it DIgital In1 or In2 nothing works. 

I probed the signals Digin1 and Digin2 using LP1 and LP2 and saw these waveforms:

Digin1 (blue In1 yellow Digin1):
![image](https://user-images.githubusercontent.com/138027/209448075-7bfacfcd-070b-44b6-8d58-096807768351.png)

and somewhat different on Digin2 (blue In2 yellow Digin2):
![image](https://user-images.githubusercontent.com/138027/209448078-5c52e5b9-c757-4d5d-b328-3b1c89f17f62.png)

The same signals then go to GPIO 14 or 15. Quite naturally the ESP32 function doesn't work. 

Here is the picture of my setup:
![image](https://user-images.githubusercontent.com/138027/209414826-360f6b70-658e-42df-af40-e70763a761c1.png)

I wonder what I'm missing. Any help would be appreciated.

**Comment by sergei** [comment]
Just to be precise I don't connect the actual B&G instruments now, but just my [HW simulator](https://github.com/sergei/sci2000/tree/master/mhusim/ftdi) based on this board: https://learn.adafruit.com/adafruit-mcp4728-i2c-quad-dac It drives the analog inputs just fine, but not the digital ones.

**Comment by sergei** [comment]
Just as a wild guess I tried to see what would happen if the diodes in the clamper are soldered with the wrong polarity:
[![CircuitLab Schematic szgue7h3yg3j](https://www.circuitlab.com/circuit/szgue7h3yg3j/screenshot/540x405/)](https://www.circuitlab.com/circuit/szgue7h3yg3j/saior-hat/)
The results seems to be very similar to what I'm seeing with the scope:
![image](https://user-images.githubusercontent.com/138027/209446506-b8263d8d-d556-4a2e-9f19-565e253d2cf3.png)

Instead of the desired behaviour:
![image](https://user-images.githubusercontent.com/138027/209446376-3c52aea4-4df8-459e-9232-b7fbe645dc24.png)

There is no way by looking at PCB I can see if they are soldered properly or not :-) 
![image](https://user-images.githubusercontent.com/138027/209446637-f915e8dd-a64b-46f6-bfa3-87bd1cf1f5c3.png)

**Comment by mairas** [comment]
Hi Sergei, I'm really sorry for my late reply!

If the top clamping diode in Digin1 would be inverted, I believe the "on" pulse wouldn't be clipped... unless it got clipped by the GPIO protection diode, of course. Other than that, your analysis would seem correct. Could you try what happens if you detach the engine hat from the SH-ESP32 and connect GND and 3.3V with some jumper wire?

Assuming you have a multimeter with a diode testing function, you could also test the clamping diodes. Cathode (black test lead) should always be towards the silk screen "U" shape bottom. Even if the part markings are too small to see, poking with a multimeter usually works. 😄

Finally, you could also double-check your code that you haven't accidentally set the input pins as INPUT_PULLUP or INPUT_PULLDOWN. That could also cause something like you observed. But if you used the demo code, that shouldn't be the case.

I have previously received a batch of proto boards, a part of which had one IC misaligned. I did check one whole panel of the 1.0.0 engine hats, and they all had the clamping diodes correctly aligned. But more crucially, I test each individual board before shipping, and all digital inputs are tested in the test jig. Even so, mistakes do happen, of course.

If the proposed tests above don't shed any light, I'll be of course happy to send you a new board. Just let me know how it goes.

**Comment by sergei** [comment]
Hi Matti,

I did remove the the engine top hat board and tested the diodes with multimeter.  The diode testing reads 0.1V forward biased, and 1.4V  reverse biased. The Ohm meter mode reads 1.2K forward biased and 14K reverse biased.  All eight of them seem behave the same.

Here is what the scope looks like (blue InF1 yellow Digin1):
![image](https://user-images.githubusercontent.com/138027/210284596-d1d9937d-b193-4ee4-9098-1ed1e720617c.png)
and the same with GND and 3.3V connected with jumper wire:
![image](https://user-images.githubusercontent.com/138027/210284632-d1a60453-ac55-4658-842f-110a5eb73f58.png)

My second board behaves the same.

**Comment by mairas** [comment]
Ahh, lol, I know realized I worded my suggestion completely wrong! I meant that you should power the board (but that of course wasn't what I wrote).

If the board is unpowered, the ICs are enough to pull Vcc to ground, meaning that the upper clamping diode will conduct and clamp the output close to zero. However, if you provide the board a normal operating voltage, then the clamping direction is as it should.

I tested myself and got identical results on an unpowered board, but when I switched the power on, the output was as it should.

**Comment by sergei** [comment]
Oh I see :-) I should have thought about it. Indeed if powered with the external supply the board works just fine:
![image](https://user-images.githubusercontent.com/138027/210397445-3f2f52b2-925d-46f8-90ca-5cca27f78560.png)
Let me check my SW then to see how the GPIOs are configured.

**Comment by sergei** [comment, marked as answer]
Yes indeed, apparently the [pcnt_unit_config()](https://docs.espressif.com/projects/esp-idf/en/v4.4.3/esp32/api-reference/peripherals/pcnt.html?highlight=pcnt_unit_config#_CPPv416pcnt_unit_configPK13pcnt_config_t) function configures the pulse input GPIO as pull up. 

I added the call:
`    gpio_set_pull_mode((gpio_num_t)pulseGpio, GPIO_FLOATING);
`
and got the desired behaviour:

![image](https://user-images.githubusercontent.com/138027/210405175-6bb8aa08-cbcb-4067-9c01-9319e53534a0.png)

**Comment by sergei** [comment]
Thanks a lot for your help Matti, I'm a happy camper now :-)

**Reply by mairas** [reply]
Great! Happy to have the issue solved!
